ivadomed/transforms.py

```diff
@@ -219,10 +221,11 @@ class Resample(ImedTransform):
         interpolation_order (int): Order of spline interpolation. Set to 0 for label data. Default=2.
     """
 
-    def __init__(self, hspace, wspace, dspace=1.):
+    def __init__(self, hspace, wspace, dspace=1., flag_pixel = False):
```

Comment:
@rezazad68 : What do you mean by "flag_pixel"? Can you please explain / document?

Reply:
It is explained in the discussion. it is a flag for resizing the sample into a given resolution rather than a fraction. https://github.com/ivadomed/ivadomed/pull/787#discussion_r637538669

Reply:
That's what I do not understand: the current implementation receives a resolution in mm, see [documentation](https://ivadomed.org/en/latest/configuration_file.html#resample).
Yes we do use `scipy.ndimage.zoom` but the transformation does not have a fraction as input: we compute the fraction required by this `scipy` function using the input image header (ie source resolution) and the transformation input `wpace`, `hspace` and `dspace` (ie destination resolution).

But maybe I'm missing something: could you please help me to understand? Cheers

Reply:
Let's say I want to resize a sample with resolution `(H*W)` to a given resolution `(H'*W')` . So, I need to calculate `aH = H'` and `bW= W'`. Where `a = H'/H`  and `b= W'/W` (line 268-270) in transforms. I am switching to this calculation mode using flag_pixel. Since the `hspace` and `wspace` are the constant float numbers in range `[0, 1]` it would create different sizes for input with variable size. If the zooms vector contains the sample resolution then the previous form calculates `a = H/H'`  and `b= W/W'` which is not suitable.

Reply:
Just to be sure: what is the unit of your H and W?
- Let's call "resolution", the size, in millimeter, of one voxel, e.g. 0.75mm x 0.75mm x 1mm --> That is currently what is implemented in `ivadomed`. Example:
```
{
    "transformation": {
        "Resample": {
            "wspace": 0.75,
            "hspace": 0.75,
            "dspace": 2
        }
    }
}
```
- Let's call "matrix size", the size, in number of voxels, of the image, e.g. 250 x 250 x 100 voxels.

However, I'm very grateful that you are raising this because: There is an error in the documentation, sorry about that:

Reply:
However, I'm very grateful that you are raising this because: There is an error in the documentation, sorry about that: `wspace` and cie are not in `[0, 1]`, they are NOT scaling factors, they are destination resolution in mm. It needs to be corrected!!

Reply:
I'm guessing that your H and W are in "number of voxels", right?
As far as I understand, in medical imaging, it is more common to resample in terms of voxel resolution rather than matrix size. For anatomical consistency reasons?
But I imagine you could achieve your goal by (i) Resample your dataset to a common voxel resolution, (ii) Crop the resulting data to a common matrix size. Would it work for your application? What are your thoughts?

Reply:
We don't have 3D data here, it is a single frame :). You can consider it as a grayscale image. That's why I used the term `flag_pixel` to highlight it. We learn and estimate the intervertebral disc location from the center slice (An average of the 6 middle slices), hence, it is a grayscale image. We don't crop the data.

Reply:
> We don't have 3D data here, it is a single frame :)

Just so you know: compatible 2D and 3D, [check this up](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/transforms.py#L54).
My question remains valid in 2D: What is the unit of your transformation input? mm or number of pix-/vox-el?

> That's why I used the term `flag_pixel` to highlight it.

Sorry, I still do not understand.
I'm happy to help but please help me to understand your point by answering the above questions.

> We don't crop the data.

I'm trying to help here. Could you please explain to me what is behind your "No"?

Why do you want to use `Resample`? My guess is: you want your the network inputs to have the same dimension (aka matrix size). Is it correct? To achieve this, my suggestion is to `Resample` all your input data to a common voxel resolution (in mm) in order to have anatomical consistency, which is quite critical for your project. Then use a `CenterCrop` to make use all your samples have the same array size.

Reply:
UPDATE: I found the answer of my question re the unit [above](https://github.com/ivadomed/ivadomed/pull/787#discussion_r632452269). Great. Now, could you share with me your thoughts re the "Resample then Crop" option?

Reply:
It might be an option but I am not sure it can affect the performance or not. The pose estimation approaches are usually uses geometrical information and it is better to have full structure. 
Do you mean Resample on data preparation level, once creating the new bids dataset?

Reply:
> uses geometrical information

That's why I'm suggesting to enforce the same pixel resolution for your data :-)

>  it is better to have full structure

Yes! I agree with you. That's why you need to choose the CropParam carefully to make sure that nothing is cropped --> it will be Padding mostly. eg take your larger image in terms of mm dimensions (in mm hein :-) ) --> Convert it to pixel dimensions after Resampling --> Take these dimensions as parameter for the Cropping. By doing so: you are certain that no anatomical structure will be cropped!

> Do you mean Resample on data preparation level, once creating the new bids dataset?

As part of the `ivadomed` pipeline, ie in the config file. No need to create a new BIDS dataset for this I think.

Reply:
Removing resize (with resample) and using crop instead affected the performance.

Reply:
> Removing resize (with resample) and using crop instead affected the performance.

Can you please elaborate? Performance in terms of quality of segmentation or speed? Also, if the use of resample was made so that it effectively replaced what you implemented previously with "resize", there should be no change. So, if there was a change then it is probably because the implementation/usage of "resample" was incorrect. More details are needed so we can help better.

Reply:
Agree with @jcohenadad !

Also, special care should be given to the Crop parameters: if too small it may crop important information --> I'd recommend to use the larger image size after resampling as Crop parameters.

Reply:
We have samples with different sizes, there are two ways to overcome this issue: 1- resize all samples to a unique size, which I have done using the Resample class with adding pixel_flag. 2- Use the original Resample and then crop, where inside the Resample we have zooms value all 1, thus, using resample has no gain. We only can use crop. Center cropping images with shapes 260*320, 320*320, and 420*400 to 256*256 loses some geometrical information, and consequently, the quality of learned intervertebral disc on the validation set is reduced.  
The modified Resample works fine but if we need to use crop (with bigger size), it might effect the performance.

Reply:
> use Center Crop instead (with a bigger size), it might affect the performance

@rezazad68: I think it is because you cropped the data with the smallest matrix size (256x256). Please see my comments above (eg [here](https://github.com/ivadomed/ivadomed/pull/787#discussion_r643561365)) --> they are suggesting you to use the larger matrix size as Cropping parameters, so in your example 420x400.

Reply:
@charleygros The image size (for example 256*320) I have is smaller than your suggested crop size. I am using [prepare_dataset_vertebral_labeling.py](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/scripts/prepare_dataset_vertebral_labeling.py) for generating the mid images. I think we need to address the resampling here because later zooms values are all 1. What do you think?

Reply:
> The image size (for example 256*320) I have is smaller than your suggested crop size.

That's not a problem: as explained above, it will just be padded. And the larger one will not be cropped.

> I think we need to address the resampling here because later zooms values are all 1. What do you think?

Sorry, I'm not sure to understand here. Could you please help me?
PS: I understand all pixel have the same resolution (what I think you call "zooms"), 1mm x 1mm

Reply:
Ok, I will update the config file with `Center Crop` instead of `resize`. 

Yes, same resolution. The zooms refer to [this](https://github.com/ivadomed/ivadomed/blob/b640728f4cf0c0a4193ea95e28d2a37285534ea7/ivadomed/transforms.py#L261) line of the code